design/034-new_lockfiles.md

```diff
@@ -0,0 +1,237 @@
+```
+
+But many things have changed, lets see them.
+
+### Non-strict vs strict
+
+If we now do a modification to our conanfile.txt to:
+
+```
+[requires]
+pkg/[>=1.0 <2.0]
+```
+
+And assuming a ``pkg/1.1`` version exist, then:
+
+```bash
+conan install . --lockfile=conan.lock  # will work
+```
+
+Will not fail by default, and will be able to resolve to ``pkg1.1``. This is because the default mode is non-strict. One of the most frustrating behaviors of Conan 1.X lockfiles was when lockfiles didn’t allow you to do modifications to your conanfiles and keep working.
+
+If we do:
+
+```bash
+conan install . --lockfile=conan.lock --lockfile-strict # will not work
+```
+
+It will throw an error saying that the requirement ``pkg/[>=1.0 <2.0]`` cannot be found in the lockfile
+
+### Multi-configuration lockfiles
+
+If we have one conanfile like this:
+
+```python
+class Pkg(ConanFile):
+    settings = "os"
+    def requirements(self):
+        if self.settings.os == "Windows":
+            self.requires("win/[>0.0]")
+        else:
+            self.requires("nix/[>0.0]")
+```
+Then it will be possible to capture the lockfile for both Windows and Linux configuration in a single lockfile file, in the following way, assuming there exist packages ``win/0.1`` and ``nix/0.1``:
+
+```bash
+conan lock create .  --lockfile-out=conan.lock -s os=Windows -s:b os=Windows
+# That captures a lockfile with ``win/0.1``
+conan lock create .  --lockfile=conan.lock --lockfile-out=conan.lock -s os=Linux -s:b os=Linux
+# That will augment the conan lock and it will end with both ``win/0.1`` and ``nix/0.1``
+```
+
+The later application of the same single resulting “conan.lock” lockfile works for both Windows and Linux configurations:
+
+```bash
+conan install . --lockfile=conan.lock -s os=Windows -s:b os=Windows
+# resolves to win/0.1, but of course nix/0.1 will not be part of the resulting graph
+
+conan install .  --lockfile=conan.lock -s os=Linux -s:b os=Linux
+# resolves to nix/0.1, but of course win/0.1 will not be part of the resulting graph
+```
+
+This approach remains valid when different configurations resolve to different versions of the same package. If the package ``conanfile.py`` is:
+
+```python
+class Pkg(ConanFile):
+    settings = "os"
+    def requirements(self):
+        if self.settings.os == "Windows":
+            self.requires("dep/0.1")
+        else:
+            self.requires("dep/0.2")
+```
+
+When the above steps for both Windows and Linux are executed, the resulting lockfile will contain both ``dep/0.1`` and ``dep/0.2``, including the latest revision for each one, but subsequent ``conan install --lockfile`` installations will still be able to resolve to the correct one, down to the locked recipe revision. This is based on the fact that the locked version still needs to match the required one, so when in Windows, it will look in the lockfile for ``dep/0.1``, will find it there, and will obtain the locked recipe revision from the lockfile.
```

Comment:
So, essentially, this looks like all different conan v1.X lockfiles are now merged together into a single giant lockfile?

Reply:
No, the Conan 1.X lockfiles were basically unmergeable, because they contained a graph, with numbered nodes for each package.

Conan 2.0 lockfiles are just lists of package references, merging them yes, is basically merging the lists and re-sorting them, in a largest lockfile.